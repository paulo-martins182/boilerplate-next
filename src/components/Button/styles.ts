'use client'

import styled, { css } from 'styled-components'
import { ButtonProps } from '.'

type ButtonStyleProps = Pick<ButtonProps, 'background'>

export const Wrapper = styled.button<ButtonStyleProps>`
  ${({ theme, background }) => css`
    all: unset;
    border-radius: 8px;
    background: ${background || theme.colors.orange.normal};
    min-height: 45px;
    padding: 0 6px;
    min-width: 100px;
    font-size: ${theme.fontSizes.medium};
    display: flex;
    align-items: center;
    justify-content: center;
    gap: ${theme.spacing.medium};
    color: ${theme.colors.white.light};
    opacity: 0.8;
    cursor: pointer;
    transition: ${theme.animationDurations.fast};
    &:hover {
      opacity: 1;
    }

    &:disabled {
      background: ${theme.colors.greyzy.normal};
      cursor: not-allowed;
    }
  `}
`
