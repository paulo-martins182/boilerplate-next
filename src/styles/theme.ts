const theme = {
  colors: {
    white: {
      light: '#FFFFFF'
    },
    blue: {
      light: '#A1B5FF',
      soft: '#7A97FF',
      normal: '#5579FF',
      dark: '#3D5FD9'
    },
    marine: {
      light: '#2E3D80',
      soft: '#253166',
      normal: '#1B244C',
      dark: '#131A36'
    },
    greyzy: {
      light: '#BEC7D4',
      soft: '#AAB2BD',
      normal: '#939AA3',
      dark: '#70767D'
    },
    orange: {
      normal: '#FF7C24'
    }
  },
  fontSizes: {
    Xlarge: '4.5rem',
    large: '4.0rem',
    mediumLarge: '2.5rem',
    medium: '1.8rem',
    small: '1.6rem',
    xsmall: '1.4rem',
    xxsmall: '1.2rem'
  },
  fontWeights: {
    light: 300,
    regular: 400,
    semibold: 600,
    bold: 700,
    xbold: 900
  },
  borderRadius: {
    small: '0.8rem',
    medium: '1.6rem',
    large: '2.4rem',
    circle: '50%'
  },
  boxShadow: {
    default: '0px 4px 8px rgba(0, 0, 0, 0.1)'
  },
  animationDurations: {
    fast: '150ms',
    normal: '300ms',
    slow: '600ms',
    verySlow: '30s'
  },
  transitions: {
    fast: '0.5s',
    normal: '1s',
    slow: '2s'
  },
  spacing: {
    xsmall: '0.4rem',
    small: '0.8rem',
    medium: '1.6rem',
    large: '2.4rem',
    xlarge: '3.2rem',
    xxlarge: '4.8rem',
    xxxlarge: '6.4rem'
  },
  gap: {
    small: '0.8rem',
    medium: '1.6rem',
    large: '2.4rem',
    xlarge: '3.2rem'
  }
}

export default theme
