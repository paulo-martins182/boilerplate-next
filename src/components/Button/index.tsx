import React, {
  AnchorHTMLAttributes,
  ButtonHTMLAttributes,
  ElementType
} from 'react'
import * as S from './styles'

type ButtonTypes =
  | AnchorHTMLAttributes<HTMLAnchorElement>
  | ButtonHTMLAttributes<HTMLButtonElement>

export type ButtonProps = {
  background?: string
  isDisabled?: boolean
  as?: ElementType
} & ButtonTypes

const Button = ({
  children,
  background,
  isDisabled,
  ...props
}: ButtonProps) => (
  <S.Wrapper background={background} disabled={isDisabled} {...props}>
    {children}
  </S.Wrapper>
)

export default Button
