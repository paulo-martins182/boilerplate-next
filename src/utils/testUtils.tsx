import theme from '@/styles/theme'
import { render, RenderOptions } from '@testing-library/react'
import { ReactElement } from 'react'
import { ThemeProvider } from 'styled-components'
import '@testing-library/jest-dom'
import 'jest-styled-components'

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export * from '@testing-library/react'

export { customRender as render }
