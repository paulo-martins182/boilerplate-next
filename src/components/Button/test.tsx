import { render, screen, fireEvent } from '@/utils/testUtils'
import Button from '.'
import theme from '@/styles/theme'

describe('Button Component', () => {
  it('should render the button and check click event', () => {
    const handleClick = jest.fn(() => console.log('clicked act'))
    render(
      <Button onClick={handleClick} as="button">
        Click Me
      </Button>
    )

    const buttonElement = screen.getByRole('button')
    expect(buttonElement).toBeInTheDocument()
    fireEvent.click(buttonElement)

    expect(handleClick).toHaveBeenCalled()
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('should render the button with correct styles when disabled', () => {
    render(
      <Button onClick={() => {}} isDisabled={true}>
        Disabled Button
      </Button>
    )

    const buttonElement = screen.getByRole('button', {
      name: /disabled button/i
    })
    expect(buttonElement).toBeDisabled()
    expect(buttonElement).toHaveStyle(
      `background-color: ${theme.colors.greyzy.normal}`
    )
    expect(buttonElement).toHaveStyle('cursor: not-allowed')
  })

  it('should render multiple children correctly', () => {
    const { getByText } = render(
      <Button>
        <span>Click</span> me
      </Button>
    )
    expect(getByText('Click')).toBeInTheDocument()
    expect(getByText('me')).toBeInTheDocument()
  })

  it('should render Button as a link', () => {
    const { getByRole, debug, container } = render(
      <Button as="a" href="#">
        Link
      </Button>
    )

    debug(container)
    expect(getByRole('link')).toBeInTheDocument()
  })

  it('should match the snapshot', () => {
    const { asFragment } = render(<Button>Snapshot Test</Button>)
    expect(asFragment()).toMatchSnapshot()
  })
})
