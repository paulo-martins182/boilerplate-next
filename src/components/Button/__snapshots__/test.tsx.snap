// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should match the snapshot 1`] = `
<DocumentFragment>
  .c0 {
  all: unset;
  border-radius: 8px;
  background: #FF7C24;
  min-height: 45px;
  padding: 0 6px;
  min-width: 100px;
  font-size: 1.8rem;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 1.6rem;
  color: #FFFFFF;
  opacity: 0.8;
  cursor: pointer;
  transition: 150ms;
}

.c0:hover {
  opacity: 1;
}

.c0:disabled {
  background: #939AA3;
  cursor: not-allowed;
}

<button
    class="c0"
  >
    Snapshot Test
  </button>
</DocumentFragment>
`;
